bitcoin/src/blockdata/script.rs

```diff
@@ -0,0 +1,215 @@
+
+//! Bitcoin scripts.
+//!
+//! *[See also the `Script` type](Script).*
+//!
+//! This module provides the structures and functions needed to support scripts.
+//!
+//! <details>
+//! <summary>What is Bitcoin script</summary>
+//!
+//! Scripts define Bitcoin's digital signature scheme: a signature is formed
+//! from a script (the second half of which is defined by a coin to be spent,
+//! and the first half provided by the spending transaction), and is valid iff
+//! the script leaves `TRUE` on the stack after being evaluated. Bitcoin's
+//! script is a stack-based assembly language similar in spirit to [Forth].
+//!
+//! Script is represented as a sequence of bytes on the wire, each byte representing an operation,
+//! or data to be pushed on the stack.
+//!
+//! See [Bitcoin Wiki: Script][wiki-script] for more information.
+//!
+//! [Forth]: https://en.wikipedia.org/wiki/Forth_(programming_language)
+//!
+//! [wiki-script]: https://en.bitcoin.it/wiki/Script
+//! </details>
+//!
+//! In this library we chose to keep the byte representation in memory and decode opcodes only when
+//! processing the script. This is similar to Rust choosing to represent strings as UTF-8-encoded
+//! bytes rather than slice of `char`s. In both cases the individual items can have different sizes
+//! and forcing them to be larger would waste memory and, in case of Bitcoin script, even some
+//! performance (forcing allocations).
+//!
+//! ## `Script` vs `ScriptBuf` vs `Builder`
+//!
+//! These are the most important types in this module and they are quite similar, so it may seem
+//! confusing what the differences are. `Script` is an unsized type much like `str` or `Path` are
+//! and `ScriptBuf` is an owned counterpart to `Script` just like `String` is an owned counterpart
+//! to `str`.
+//!
+//! However it is common to construct an owned script and then pass it around. For this case a
+//! builder API is more convenient. To support this we provide `Builder` type which is very similar
+//! to `ScriptBuf` but its methods take `self` instead of `&mut self` and return `Self`. It also
+//! contains a cache that may make some modifications faster. This cache is usually not needed
+//! outside of creating the script.
+//!
+//! At the time of writing there's only one operation using the cache - `push_verify`, so the cache
+//! is minimal but we may extend it in the future if needed.
+
+#[cfg(test)]
+mod tests;
+
+use io::{BufRead, Write};
+use primitives::opcodes::all::*;
+
+use crate::consensus::{encode, Decodable, Encodable};
+use crate::internal_macros::define_extension_trait;
+use crate::key::WPubkeyHash;
+use crate::policy::DUST_RELAY_TX_FEE;
+use crate::prelude::sink;
+use crate::taproot::{LeafVersion, TapLeafHash};
+use crate::FeeRate;
+
+/// Re-export everything from the [`primitives::script`] module.
+#[rustfmt::skip]                // Keep public re-exports separate.
+#[doc(inline)]
+pub use primitives::script::*;
+
+define_extension_trait! {
```

Comment:
In b461457c50ec7aaf9866b267c8c607edefa5dcbc:

Can you move the extension-trait-ification of these methods to its own commit? It has far-reaching implications and git doesn't recognize it as a code move.

Reply:
This I can definitely do, thanks.

Reply:
Just flagging that I don't know what you mean by far reaching implications, since we already introduced the strategy of extension traits.

As commented below, I had a good god at doing as you suggest and it was not creating patches any easier to review (IMO). Please do stick to your guns if I am missing something.

Reply:
I deleted the message I wrote below, will have another go.

Reply:
I just mean it pollutes the diff.